### ISSUE TITLE
History don't work with always transitions

### ISSUE THREAD

**Issue by elyzov** [statelyai/xstate#1685]
**Description**
I have a machine which represent deploying of some product.

```
const machine = Machine({
  id: "activity",
  initial: "idle",
  states: {
    idle: {
      id: "idle",
      initial: "absent",
      states: {
        absent: {
          on: {
            DEPLOY: "#deploy"
          }
        },
        present: {
          on: {
            DEPLOY: "#deploy",
            DESTROY: "#destroy"
          }
        },
        hist: {
          type: "history"
        }
      }
    },
    deploy: {
      id: "deploy",
      on: {
        SUCCESS: "#idle.present",
        FAILURE: "#idle.hist"
      }
    },
    destroy: {
      id: "destroy",
      always: [
        { target: '#idle.absent' },
      ],
    }
  }
});
```

I've noticed the following path don't work as expected:

```
[idle.absent] -> DEPLOY -> [deploy] -> SUCCESS -> [idle.present] -> DESTROY -> [idle.absent] -> DEPLOY -> [deploy] -> FAILURE -> ?
```

**Expected Result**
I expect the machine will be return to the `[idle.absent]` state.

**Actual Result**
Machine returns to the `[idle.present]` state.

**Reproduction**

- ⚛︎ [CodeSandbox template (React)](https://codesandbox.io/s/xstate-react-history-with-always-yug63)

**Additional context**
XState version - 4.14.1

**Comment by davidkpiano** [comment]
There is some confusion here around IDs. When you specify a target, it can (currently) either be:

- A full ID (`#someIdentifier`)
- A relative state (`sibling` or `.child.grandchild` or `sibling.nibling`)

You can't mix both of them:

- ❌ `#someIdentifier.descendant`

In this case, you can target `"idle.hist"` from the `FAILURE` transition instead of `"#idle.hist"`.

**Comment by elyzov** [comment]
I've changed references, but got the same result: https://codesandbox.io/s/xstate-react-history-with-always-yug63

**Comment by tomByrer** [comment]
I think there is a more general issue with 'always' not triggering event changes.  I get nothing in `.onTransition` if I use `always: nextstep`, but `after: { 1: 'nextstep' }` does.

Thought about opening another issue (I still can), but 95% certain these are related.
			},

**Comment by Andarist** [comment]
Please always try to share a repro case in a runnable form - either by providing a git repository to clone or a codesandbox. OSS maintainers usually can't afford the time to set up the repro, even if exact steps are given. This would help us to understand your issue better.

**Comment by tomByrer** [comment]
demo, sorry if not minimal, need to ship before Christmas ;)
https://tombyrer.github.io/cat-christmas/cat-christmas

`increment & route` do not log due to `always`
`play & end` do log because of `after` timed delay.  I needed for end to log, so I had to add
```js
after: { 1: 'increment', }
```

**Comment by Andarist** [comment]
Ok, so it looks like your issue comes from an expectation that `onTransition` should fire after each transition, including always transitions as those are transitions as well, after all. But actually `onTransition` only after the last transition in a macro step (after machine "settles" that macrostep which includes resolving all always transitions and transitions caused by raised events). This is IMHO a reasonable expectation and it's something I've already wanted to raise as a concern for the future v5 but didn't got to it yet. Could you create a new issue about this one so we could track the discussion about it there?

**Comment by davidkpiano** [comment]
The original issue works with the latest versions of `xstate` and `@xstate/react`: https://codesandbox.io/s/xstate-react-history-with-always-forked-kud5qk
